### ISSUE TITLE
Compile error but success when doc-testing a crate that has the same name as a module

### ISSUE THREAD

**Issue by HeroicKatora** [rust-lang/rust#68488]
In a crate that has the same name as a top-level module, exporting a name from this module causes rustdoc to print an error when it is ran. The command still reports an exit status of `0`. This reproduces on latest nightly and stable.

```rust
// Crate name: `this_crate`.
mod this_crate { pub struct Item; }

pub use this_crate::Item;
```

<details>
<summary>error[E0659]: `this_crate` is ambiguous</summary>
<pre>
> cargo test --doc
    Finished test [unoptimized + debuginfo] target(s) in 0.00s
   Doc-tests this_crate
error[E0659]: `this_crate` is ambiguous (name vs any other name during import resolution)
 --> /tmp/test_dir/no-alloc/src/lib.rs:7:9
  |
7 | pub use this_crate::Item;
  |         ^^^^^^^^^^ ambiguous name
  |
  = note: `this_crate` could refer to a crate passed with `--extern`
  = help: use `::this_crate` to refer to this crate unambiguously
note: `this_crate` could also refer to the module defined here
 --> /tmp/test_dir/no-alloc/src/lib.rs:1:1
  |
1 | mod this_crate { pub struct Item; }
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = help: use `crate::this_crate` to refer to this module unambiguously

error: aborting due to previous error

For more information about this error, try `rustc --explain E0659`.

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
&gt; echo $?
0
</pre>
</details>

**Comment by ollie27** [comment]
This should have been fixed recently by #68357. It errors correctly for me using `rustdoc 1.42.0-nightly (41f41b235 2020-01-23)`. Does it still reproduce for you with the latest nightly?

**Comment by HeroicKatora** [comment]
@ollie27  The latest nightly errors, so I should probably change the title.

You mean the correct behaviour is erroring? I did not realize this. It seems quite surprising to that the crate itself compiles (`cargo build`) but then would not succeed in doc tests due to code that is not actually within a doctest.

**Comment by ollie27** [comment]
Yeah, actually that error does look wrong. cargo seems to pass `--extern this_crate=<path to libthis_crate.rlib>` to `rustdoc --test` which rustdoc either needs to ignore or shouldn't be passed at all.

**Comment by joshlf** [comment]
Chiming in to say that this is affecting us in zerocopy now (though we can probably work around it).

**Comment by LGFae** [comment]
I have also run into this issue just now.

Is there an easy way of disabling `rustdoc` when running `cargo test`?
